### PR TITLE
Fix onClose event handler for Blocks

### PR DESCRIPTION
**Pull request by mikejolley**
Specifically, in Block Checkout, the `onClose` handler was not being fired when dismissing the payment request modal. I believe this was in part due to how event handlers were removed when a hook is unmounted using `removeAllListeners` which removes all events. Since we have multiple `useEffects` all at once, I think there was some kind of race condition.

I've fixed this by removing specific events on unmount rather than all of them.

# Testing instructions

On block checkout, pay using the Stripe Express Checkout but cancel the payment request. onClose handler should fire. This is easier to verify when using https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4228 since the loading state of the express payment methods will change once closed. Without this fix, the loading state will be stuck.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

File: client/blocks/payment-request/hooks.js
Comment (dechov) [inline client/blocks/payment-request/hooks.js]: The naming here sounds like it would actually return/yield a payment request, though I'm not sure what a better name might be.

File: client/blocks/payment-request/hooks.js
Comment (dechov) [inline client/blocks/payment-request/hooks.js]: Do you know if it's still necessary to set this to `null` first, or could this be done in a `! result` else case above?

File: client/blocks/payment-request/hooks.js
Comment (dechov) [inline client/blocks/payment-request/hooks.js]: I'm struggling to understand why the `noop` is necessary or how its method (stub) would ever run, since the conditions for the assignment above and the `removeEventListener` call below are the same. I see a similar thing in the Blocks extension's built-in integration ([here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/af688c2d710d405c09d555bffba90b9a3966b1a8/assets/js/payment-method-extensions/payment-methods/stripe/payment-request/use-initialization.js#L136)), but it looks like the condition was different when it was first introduced ([here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2484/files#diff-39e15c37c322b17e451f21e00c9f83311ddf00ae49f40f4653befa2805243e5cR216)).

**Review by dechov — COMMENTED** [review]
The fix consists of removing listeners for the _specific_ events instead of _all_, and the rest of the changes is basically refactoring, is that correct?

That basic change LGTM. Can confirm that the `cancel` listener would previously have been removed by the time the event fired – logging the `source` and `cancel` handler-related lines, here's what's I see on every single page load (with nothing else before opening + closing the payment request):

```
adding source event handler
adding cancel event handler
removing source event handler
adding source event handler
```

Not sure why it's adding, removing, and then readding the source, but it might have to do with the setting of `paymentRequestType`, since the `useCancelHandler` hook doesn't depend on it but the rest do.

As for the other changes, I left just a few questions.

File: client/blocks/payment-request/hooks.js
Comment (mikejolley) [reply client/blocks/payment-request/hooks.js]: The `noop` is an alternative to the optional chaining and ensures the `removeEventListener` always exists, even if these conditions change in the future, or if the unmount function runs for whatever reason before the handler is set. The noop in the original blocks was for `removeAllListeners`, but we don't use that here.

File: client/blocks/payment-request/hooks.js
Comment (mikejolley) [reply client/blocks/payment-request/hooks.js]: Happy to change if you have suggestions. This function is only used within this `useEffect` hook so is not in the public facing API.

File: client/blocks/payment-request/hooks.js
Comment (mikejolley) [reply client/blocks/payment-request/hooks.js]: I don't know, I just re-implemented what was here before. It will be null inbetween the payment request being invalidated then recreated.

**Comment by mikejolley** [comment]
> The fix consists of removing listeners for the specific events instead of all, and the rest of the changes is basically refactoring, is that correct?

Yes. When this code was refactored and moved to this repo, it seems the single `useEffect` was converted to multiple `useEffects` which broke events.

File: client/blocks/payment-request/hooks.js
Comment (reykjalin) [reply client/blocks/payment-request/hooks.js]: Yes, the `null` is necessary so we can recreate the the payment request when the shipping needs in the cart change.

For example; you have a product that needs shipping and a product that doesn't need shipping in your cart. Without this reset the payment request will insist on a shipping address even if you remove the product that needs shipping from the cart.

This happens because we can't update the `needsShipping` prop on the payment request; we must create a new payment request to update the value.

File: client/blocks/payment-request/hooks.js
Comment (reykjalin) [inline client/blocks/payment-request/hooks.js]: Out of curiosity; is there a reason you prefer to use a combination of `let`, `if`, and `noop` instead of the optional chaining?

I ask because I switched to the optional chaining because (IMO) the code was easier to understand when using optional chaining, and seeing you add it back into the code makes me wonder if there was a reason for using the "`noop`" approach in the first place 🤔 

If there are no architectural — or otherwise engineering related — differences between the two approaches, I think I'd prefer to keep the optional chaining for readability.

File: client/blocks/payment-request/hooks.js
Comment (dechov) [reply client/blocks/payment-request/hooks.js]: With @mikejolley's change here, I think it will rerun the one effect (and so still recreate the payment request) when those same dependencies change. Because of that, It doesn't look to me like setting it to `null` is necessary anymore – except to reset it in the logic above, if `! canMakePayment`.

File: client/blocks/payment-request/hooks.js
Comment (dechov) [reply client/blocks/payment-request/hooks.js]: The cleanup function isn't created until after the handler is assigned, but if we'd like to allow for the condition to change in the future, that sounds fine to me :+1:

In that case, I don't see the need for this 2nd `if` statement, since it's already been ensured that the `removeEventListener` will exist, as you point out. Not a big deal, since it's harmless to include it, but it would be nice to keep the code a bit tighter (unless I'm missing something about why we need it)

File: client/blocks/payment-request/hooks.js
Comment (reykjalin) [reply client/blocks/payment-request/hooks.js]: Ah, I didn't notice that.

I initially split them up because we only want to invalidate the payment request when specific variables change, and it's not necessary to recreate the whole payment request every time — which is quite expensive because of the network request involved to get the cart.

I haven't looked into whether joining the 2 effects increases the number of times we reset the payment request, but I think we should verify that before joining them.

Aside from that, I think you're right Paul; we can remove the `setPaymentRequest( null )` call if we decide to join the effects like this 🙂

File: client/blocks/payment-request/hooks.js
Comment (mikejolley) [reply client/blocks/payment-request/hooks.js]: I am not opinionated about this in any way, I was just basing my approach on the implementation in Blocks since the event handler implementation in this extension is broken. Personally, I see no difference in readability. If anything, the `noop` tells you what to expect in the object so really is no different. 

@nerrad may have more thoughts on this.

**Comment by mikejolley** [comment]
I think you all get the gist of the problem and fix so I'm unassigning myself from this PR. Please feel free to cherry-pick from this or make whatever changes you deem necessary to get it shipped :)

File: client/blocks/payment-request/hooks.js
Comment (nerrad) [reply client/blocks/payment-request/hooks.js]: "Readability" can sometimes be very subjective. There are times when I find optional chaining to be harder to follow :). There's definitely no architectural differences between the two approaches - mostly a preference thing.

File: client/blocks/payment-request/hooks.js
Comment (dechov) [reply client/blocks/payment-request/hooks.js]: Kept the `noop` and removed the second `if` statement in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1603/commits/ed579b6d5d463e13b5806986990a1e2bd99c6819

File: client/blocks/payment-request/hooks.js
Comment (dechov) [reply client/blocks/payment-request/hooks.js]: AFAICT it's functionality identical and should reevaluate the effect in the same circumstances as it did before. I've moved the reset call in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1603/commits/29cb4e470aa7e6570458046177672aa92ea2d95e

File: client/blocks/payment-request/hooks.js
Comment (dechov) [reply client/blocks/payment-request/hooks.js]: Changed to `createPaymentRequest` in https://github.com/woocommerce/woocommerce-gateway-stripe/commit/29cb4e470aa7e6570458046177672aa92ea2d95e

**Comment by dechov** [comment]
I've pushed a couple of minor changes – will get a final review from @reykjalin or someone else on @woocommerce/fractal before merging.

File: client/blocks/payment-request/hooks.js
Comment (reykjalin) [inline client/blocks/payment-request/hooks.js]: I'd still _prefer_ to not use a `let` binding at all. It's nitpick-y in small functions like these, but I'd still prefer to use `const` and optional chaining instead of `let` and `if`:

```js
useEffect( () => {
    const handler = paymentRequest?.on(
        'shippingaddresschange',
        shippingAddressChangeHandler( paymentRequestType )
    );

    return () => {
        handler?.removeEventListener( 'shippingaddresschange' );
    };
}, [ paymentRequest, paymentRequestType ] );
```

There's no need to reason about whether `paymentRequest` is null or not — it's automatically handled by the optional chaining so the additional overhead of parsing the `if` statement is removed.

That said, I don't think it's a required change to merge. Just a strong preference on my part 🙂.

**Review by reykjalin — COMMENTED** [review]
I still have some reservations about the `let` + `if` combo, but otherwise this looks good and tests well.

I'm holding off on an approval until @woocommerce/fractal has had a chance to discuss that a bit more. I'll bring it up in Slack so we can discuss 👍

File: client/blocks/payment-request/hooks.js
Comment (reykjalin) [reply client/blocks/payment-request/hooks.js]: Slack discussion started in p1624477548460700-slack-C01BZUL57SQ

Comment (dechov) [reply client/blocks/payment-request/hooks.js]: Sounds good to me – done in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1603/commits/d82f0e2c398e218385fab76a3ea26383634dfd01 👍

**Comment by dechov** [comment]
@reykjalin Thanks – I've updated it and rebased.

**Comment by reykjalin** [comment]
@dechov I think you accidentally removed the optional chaining again when you fixed the rebasing mistake 🤔 
Would you restoring the optional chaining logic again? 😬

**Comment by dechov** [comment]
@reykjalin Whoops – I've removed that commit and just re-did the rebase.

**Review by reykjalin — APPROVED** [review]
LGTM 🚀

**Pull request merged**
